### PR TITLE
Add code to serialize PETSc vectors.

### PR DESCRIPTION
**Pull request by drwells**
This is more flexible than `SolutionTransfer`.

One possible change would be only permitting non-ghosted vectors to be serialized - should we be that picky? I'm not sure if that choice fits our normal notion of ghosted vectors or not.

### CODE REVIEW

**Review by kronbichler — APPROVED** [review]

**Comment by drwells** [comment]
I forgot to add my test - fixed!

File: include/deal.II/lac/petsc_vector_base.h [hunk 1]
```diff
@@ -1271,6 +1308,66 @@ namespace PETScWrappers
...
+    // forward to serialization function in the base class.
+    ar &static_cast<const Subscriptor &>(*this);
+    ar &size();
+    ar &local_range();
+
+    const PetscScalar *array = nullptr;
+    int                ierr  = VecGetArrayRead(*this, &array);
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+    boost::serialization::array_wrapper<const double> velocity_wrapper(
```
Comment (tjhei) [inline include/deal.II/lac/petsc_vector_base.h:1324]: Maybe a more generic name is more appropriate 😉

File: include/deal.II/lac/petsc_vector_base.h [hunk 2]
```diff
@@ -1271,6 +1308,66 @@ namespace PETScWrappers
...
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+  }
+
+
+
+  template <class Archive>
+  inline void
+  VectorBase::load(Archive &ar, const unsigned int)
+  {
+    // the load stuff again from the archive
```
Comment (tjhei) [inline include/deal.II/lac/petsc_vector_base.h:1338]:
```suggestion
```

**Review by tjhei — APPROVED** [review]

File: include/deal.II/lac/petsc_vector_base.h (hunk 1 above)
Comment (drwells) [reply include/deal.II/lac/petsc_vector_base.h]: Yes, this is worth fixing.

**Pull request merged**